server/src/main/java/org/elasticsearch/common/util/ReleasableIntArray.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+class ReleasableIntArray implements IntArray {
+    private ReleasableBytesReference ref;
+
+    ReleasableIntArray(StreamInput in) throws IOException {
+        ref = in.readReleasableBytesReference();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBytesReference(ref);
+    }
+
+    @Override
+    public long size() {
+        return ref.length() / 4;
+    }
+
+    @Override
+    public int get(long index) {
+        if (index > Integer.MAX_VALUE / 4) {
+            throw new UnsupportedOperationException(); // NOCOMMIT Oh god, what do we do here?
```

Comment:
Here in this particular line of code it's probably a non-issue since we only read this from the wire and we'll never run into a case where this kind of index isn't out of bounds. Maybe just throw an out of bounds exception here?